### PR TITLE
[dir-707] instance detail page

### PR DESCRIPTION
**Pull request by stefan-kracht**
### [Preview 🚀](https://direktiv-ui-refs-pull-426-merge.direktiv.dev)

CleanShot 2023-08-04 at 20 02 06@2x

https://github.com/direktiv/direktiv-ui/assets/121789579/dda2fb62-7488-4280-b488-2be87eb451d9

## scope and notes
- Implementation of the instance page
- first implementation of data streaming. This is the general implementation concept (based on [this article](https://tkdodo.eu/blog/using-web-sockets-with-react-query) from one of the maintainers of ReactQuery).
- the layout is responsive (not perfect, but it has a good foundation)
- Mayor changes to the original layout
  - Introducing a header that shows all information that are shown on the instance list page
  - merging the input and output panel
    - logs get more vertical space
    - you probably don't need to see the input and output at the same time that often
    - when the instance is running, we have more screen space to read the logs and still see the input (the output is empty at that state anyway)
    - when the instance is finished, the panel will automatically switch to the output tab, and you can see the output of the instance instantly
    - in any case, you can always maximize any panel
  - the child instances are now shown in the header instead of the diagram. It will only show the latest 50 for now, to avoid pagination. As soon as there are more than 50 instances, it will show it to the user)
- I used [TanStack Virtual](https://tanstack.com/virtual/v3) to virtualize the logs. It's still in beta, but I found other packages very unintuitive in the past and I since I trust TanStack a lot (they make ReactQuery!) I was ready to give it a try. And it works super well.
- the old logs offered a button to toggle word wrap, this feature is removed for now. It's pretty complicated because the logs are virtualized and horizontal scrolling might need a second instance of that virtualizer. (the linewrap button, and state management is already implemented, but commented out). I think this feature is not that urgent to be a reason to delay our (I also find it more important to not have scroll horizontally)
- when you minimize/maximize a panel or toggle verbose logs, this will now be stored in local storage. I think this is a user preference that is worth storing for the entire "session". I could remember a user debugging a workflow on a small screen and that just want to see verbose logs on a maximized panel all the time. 
- Badges can now have an icon (the pending icon is animated). I used this on the instance page and update the instance list as well CleanShot 2023-08-04 at 14 32 27@2x
- Every log entry from our db has a log level from the type `"debug", "info", "error", "panic"`. `error` and `panic` will be displayed in red, info in blue and `debug` with no background color
- I used React Context to avoid prop drilling the instance id and filters. I wanted to go for zustand, but the store should not be global, and the id should be instantiated via a prop (to make it mandatory). That was surprisingly [much boilerplate](https://docs.pmnd.rs/zustand/guides/initialize-state-with-props) that I decided against it.
-  not 100 % happy with the header design and responsiveness, but I think it is good for now 

What and how to test:

I [prepared a ticket](https://linear.app/direktiv/issue/DIR-719/instance-page) for future e2e tests. Reading this ticket should give you a good overview about all the features on that page.

The namespace `stefan` has some handy workflows that you can use for testing:
- `delay.yaml` runs a simple loop with a timer step in between and will succeed eventually
- `stable-diffusion.yaml` is a workflow that runs very long, produces a lot of logs and will fail eventually

Issues I created that are not mandatory and can be done separately (maybe post release)

- [add tooltips to the time on the instance page](https://linear.app/direktiv/issue/DIR-720/add-tooltip-to-instance-page) 
- [input/output tab selection changes accidentally](https://linear.app/direktiv/issue/DIR-721/inputoutput-tab-selection-changes-accidentally)
- [add word wrap to the instance logs](https://linear.app/direktiv/issue/DIR-723/add-word-wrap-to-the-instance-logs)
- [minor TS issue on FilterObj](https://linear.app/direktiv/issue/DIR-724/minor-ts-issue-on-filterobj)
- [fix scroll to bottom monkey patch](https://linear.app/direktiv/issue/DIR-725/fix-scroll-to-bottom-monkey-patch)

### CODE REVIEW

File: src/pages/namespace/Instances/Detail/store/instanceContext.tsx [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    throw new Error("useFilter must be used within a InstanceContext");
+  }
+
+  return context.instanceId;
+};
+
+const useFilters = () => {
+  const context = useContext(InstanceStateContext);
+  if (!context) {
+    throw new Error("useFilter must be used within a InstanceContext");
```
Comment (sebxian) [inline src/pages/namespace/Instances/Detail/store/instanceContext.tsx:73]:
```suggestion
    throw new Error("useFilters must be used within a InstanceContext");
```
If I understand correctly this should match the method name?

**Review by sebxian — CHANGES_REQUESTED** [review]
Amazing work, @stefan-kracht, the instances detail page looks fantastic and works well as far as I have been able to test.

For time reasons, I could not read every line of code thoroughly, but in general, your code is very well structured, good to understand and you added comments in just the right places to explain newly introduced patterns. I also liked the pragmatic decisions you took regarding some problems by creating issues for later.

Regarding word wrap in the log: I honestly think the current behavior (which wraps long lines) is good enough and we should discuss with the rest of the team whether adding more complexity by adding an option to turn it off is justified. IMO, we should only do that if our users insist this is a must have feature.

There are some minor comments in the code on possible improvements.

Some of these are regarding capitalization of UI strings. I noticed we are a bit inconsistent in this regard, but I only addressed one pattern that I found quite obviously "wrong", which is the capitalization of multiple words in a sentence (e.g., "Child Instances"). At some later point, we could return to this and discuss whether we always want to start sentences (or, every ui string) with a capital letter, which seems to be the norm elsewhere. But this is nothing we should spend too much time on right now.

File: src/pages/namespace/Instances/Detail/Header/index..tsx [hunk 2]
```diff
@@ -0,0 +1,82 @@
+import { Box, FileSymlink } from "lucide-react";
```
[the remark this answers is not in the document]
Comment (stefan-kracht) [reply src/pages/namespace/Instances/Detail/Header/index..tsx]: fixed in 59f96db1754bb2c37f0ff5075bfdf388f0ff7257

File: src/api/logs/query/get.ts [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+            newResults = msg.results.filter((entry) =>
+              newestLogTimeFromCache.isBefore(entry.t)
+            );
+          }
+
+          return {
+            ...old,
+            results: [...old.results, ...newResults],
+          };
+        }
```
[the remark this answers is not in the document]
Comment (stefan-kracht) [reply src/api/logs/query/get.ts]: That is a very good idea (7f27ffc34b73b02003df39e21df9317b22cfff03)

File: src/pages/namespace/Instances/Detail/store/instanceContext.tsx (hunk 1 above)
Comment (stefan-kracht) [reply src/pages/namespace/Instances/Detail/store/instanceContext.tsx]: That's right, I updated it for both methods (ad5fefdf02bb153abb8e7d140f4a9b99b764867a)

File: src/pages/namespace/Instances/Detail/Main/Logs/Filters/TextInput.tsx [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+}: {
+  field: FilterField;
+  value?: string;
+  setFilter: (field: FilterField, value: string) => void;
+  clearFilter: (field: FilterField) => void;
+}) => {
+  const [inputValue, setInputValue] = useState<string>(value || "");
+  const { t } = useTranslation();
+
+  const handleKeyDown = (event: { key: string }) => {
```
[the remark this answers is not in the document]
Comment (stefan-kracht) [reply src/pages/namespace/Instances/Detail/Main/Logs/Filters/TextInput.tsx:22]: Thanks, I assigned the issue to me. I will take care of it, when both branches are merged

File: src/api/logs/query/get.ts [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+  stateName?: string;
+};
+
+export const getFilterQuery = (filters: FiltersObj = {}) => {
+  let query = "";
+  const workflowName = filters?.workflowName ?? "";
+  const stateName = filters?.stateName ?? "";
+  if (workflowName || stateName) {
+    query = query.concat(
+      `&filter.field=QUERY&filter.type=MATCH&filter.val=${workflowName}::${stateName}::`
```
[the remark this answers is not in the document]
Comment (stefan-kracht) [reply src/api/logs/query/get.ts]: It actually works! I would love to use CONTAINS over MATCH. It really bothered me that I always have to enter it exactly right (39b099f039b4f58b0da0e46f4c63a1315c0fcad5)

**Comment by stefan-kracht** [comment]
Thanks for the review @sebxian. You found some very valuable improvements. I should have corrected all of this by now

**Review by sebxian — APPROVED** [review]
I tested the filters again after the reverted changes and they work as expected.

All looks great now, ready to merge :-)

**Comment by stefan-kracht** [comment]
Just added [one last commit](https://github.com/direktiv/direktiv-ui/pull/426/commits/7d4bfaa7c299f8d98f04aeb29be468f84daf2e21) that implements the submit button to the filters ([DIR-729](https://linear.app/direktiv/issue/DIR-729/add-submit-buttons-to-filters-in-instance-detail-log)).

**Pull request merged**